source/common/http/alternate_protocols_cache_impl.cc

```diff
@@ -2,22 +2,84 @@
 
 #include "source/common/common/logger.h"
 
+#include "quiche/spdy/core/spdy_alt_svc_wire_format.h"
+
 namespace Envoy {
 namespace Http {
+namespace {
+std::string originToString(const AlternateProtocolsCache::Origin& origin) {
+  return absl::StrCat(origin.scheme_, "://", origin.hostname_, ":", origin.port_);
+}
+} // namespace
+
+std::string AlternateProtocolsCacheImpl::protocolsToStringForCache(
+    const std::vector<AlternateProtocol>& protocols, TimeSource& /*time_source*/) {
+  if (protocols.empty()) {
+    return std::string("clear");
+  }
+  std::string value;
+  for (auto& protocol : protocols) {
+    if (!value.empty()) {
+      value.push_back(',');
+    }
+    absl::StrAppend(&value, protocol.alpn_, "=\"", protocol.hostname_, ":", protocol.port_, "\"");
+
+    // Note this is _not_ actually the max age, but the absolute time at which
+    // this entry will expire. protocolsFromString will convert back to ma.
+    absl::StrAppend(
+        &value, "; ma=",
+        std::chrono::duration_cast<std::chrono::seconds>(protocol.expiration_.time_since_epoch())
+            .count());
+  }
+  return value;
+}
 
-AlternateProtocolsCacheImpl::AlternateProtocolsCacheImpl(TimeSource& time_source)
-    : time_source_(time_source) {}
+absl::optional<std::vector<AlternateProtocolsCache::AlternateProtocol>>
+AlternateProtocolsCacheImpl::protocolsFromString(absl::string_view alt_svc_string,
+                                                 TimeSource& time_source, bool from_cache) {
+  std::vector<AlternateProtocol> protocols;
+  spdy::SpdyAltSvcWireFormat::AlternativeServiceVector altsvc_vector;
+  if (!spdy::SpdyAltSvcWireFormat::ParseHeaderFieldValue(alt_svc_string, &altsvc_vector)) {
+    return {};
+  }
+  for (const auto& alt_svc : altsvc_vector) {
+    MonotonicTime expiration;
+    if (from_cache) {
+      auto expire_time_from_epoch = std::chrono::seconds(alt_svc.max_age);
+      auto time_since_epoch = std::chrono::duration_cast<std::chrono::seconds>(
+          time_source.monotonicTime().time_since_epoch());
+      if (expire_time_from_epoch < time_since_epoch) {
+        expiration = time_source.monotonicTime();
```

Comment:
This seems to set the expiration to the current time when an expired cache entry is parsed. Below in `AlternateProtocolsCacheImpl::findAlternatives` we only ignore entries when the expiration is earlier than the current time  (`now > protocol.expiration_`) .

I don't see where this PR actual retrieves values from the cache, but it seems that retrieving an expired entry might cause it to be considered unexpired, if only briefly? Maybe the resolution of monotonicTime() is always high enough that it's unlikely to ever actually happen, but I gather the resolution is system-defined.

Reply:
yeah largely if it's expired we set the expiration time now.  if there were jitter, it might end up loaded temporarily but there's always going to be jitter with these things so I think I'm largely not concerned.